jwst/photom/photom.py

```diff
@@ -643,25 +636,47 @@ def photom_io(self, tabdata):
             waves = tabdata['wavelength'][:nelem]
             relresps = tabdata['relresponse'][:nelem]
 
+            # Make sure waves and relresps are in increasing wavelength order
+            if not np.all(np.diff(waves) > 0):
+                index = np.argsort(waves)
+                waves = waves[index].copy()
+                relresps = relresps[index].copy()
+
             # Convert wavelengths from meters to microns, if necessary
             microns_100 = 1.e-4         # 100 microns, in meters
             if waves.max() > 0. and waves.max() < microns_100:
                 waves *= 1.e+6
             wl_unit = 'um'
 
-            # Set the relative sensitivity table for the correct Model type
-            log.info('Storing relative response table')
+            # Compute a 2-D grid of conversion factors, as a function of wavelength
             if isinstance(self.input, datamodels.MultiSlitModel):
-                otab = np.array(list(zip(waves, relresps)),
-                       dtype=self.input.slits[self.slitnum].relsens.dtype)
-                self.input.slits[self.slitnum].relsens = otab
-                self.input.slits[self.slitnum].relsens.columns['wavelength'].unit = wl_unit
-
+                wl_array = expand_to_2d.get_wavelengths(self.input.slits[self.slitnum])
             else:
-                otab = np.array(list(zip(waves, relresps)),
-                                dtype=self.input.relsens.dtype)
-                self.input.relsens = otab
-                self.input.relsens.columns['wavelength'].unit= wl_unit
+                wl_array = expand_to_2d.get_wavelengths(self.input)
+
+            wl_array[np.isnan(wl_array)] = -1.
+            conv_2d = np.interp(wl_array, waves, relresps, left=0., right=0.)
+
+            # Combine the scalar and 2-D conversions
+            # NOTE: the 2-D conversion is divided into the data for now, until the
+            # instrument teams deliver multiplicative conversions in photom ref files
+            conversion = conversion / conv_2d
+
+        # Apply the conversion to the data and all uncertainty arrays
+        if isinstance(self.input, datamodels.MultiSlitModel):
+            self.input.slits[self.slitnum].data *= conversion
+            self.input.slits[self.slitnum].err *= conversion
+            self.input.slits[self.slitnum].var_poisson *= conversion
```

Comment:
Shouldn't the variances be multiplied by `conversion**2`?

Reply:
I wasn't sure about that. This is essentially a units change that's being applied, not a correction of some sort.

Reply:
You're right. Karl confirms it should be multiplied by the square of the conversion.